src/hydranfc/hydranfc_bbio_reader.c

```diff
@@ -0,0 +1,146 @@
+
+    /* Test ISO14443-A/Mifare read UID */
+    Trf797xInitialSettings();
+    Trf797xResetFIFO();
+
+    /*
+     * Write Modulator and SYS_CLK Control Register (0x09) (13.56Mhz SYS_CLK
+     * and default Clock 13.56Mhz))
+     */
+    data_buf[0] = MODULATOR_CONTROL;
+    data_buf[1] = 0x31;
+    Trf797xWriteSingle(data_buf, 2);
+
+    /*
+     * Configure Mode ISO Control Register (0x01) to 0x88 (ISO14443A RX bit
+     * rate, 106 kbps) and no RX CRC (CRC is not present in the response))
+     */
+    data_buf[0] = ISO_CONTROL;
+    data_buf[1] = 0x88;
+    Trf797xWriteSingle(data_buf, 2);
+
+    data_buf[0] = ISO_CONTROL;
+    Trf797xReadSingle(data_buf, 1);
+}
+
+void set_mode_iso_15693(void) {
+    static uint8_t data_buf[2];
+
+    /* Test ISO15693 read UID */
+    Trf797xInitialSettings();
+    Trf797xResetFIFO();
+
+    /* Write Modulator and SYS_CLK Control Register (0x09) (13.56Mhz SYS_CLK and default Clock 13.56Mhz)) */
+    data_buf[0] = MODULATOR_CONTROL;
+    data_buf[1] = 0x31;
+    Trf797xWriteSingle(data_buf, 2);
+
+    /* Configure Mode ISO Control Register (0x01) to 0x02 (ISO15693 high bit rate, one subcarrier, 1 out of 4) */
+    data_buf[0] = ISO_CONTROL;
+    data_buf[1] = 0x02;
+    Trf797xWriteSingle(data_buf, 2);
+}
+
+
+void bbio_mode_hydranfc_reader(t_hydra_console *con) {
+    uint8_t bbio_subcommand;
+    uint8_t rlen, clen;
+    uint8_t *rx_data = (uint8_t *) g_sbuf + 4096;
+
+    init_gpio(con);
+
+    bbio_mode_id(con);
+
+    while (!hydrabus_ubtn()) {
+
+        if (chnRead(con->sdu, &bbio_subcommand, 1) == 1) {
+            switch (bbio_subcommand) {
+                case BBIO_NFC_SET_MODE_ISO_14443A: {
+                    set_mode_iso_14443A();
+                    break;
+                }
+                case BBIO_NFC_SET_MODE_ISO_15693: {
+                    set_mode_iso_15693();
+                    break;
+                }
+                case BBIO_NFC_RF_OFF: {
+                    Trf797xTurnRfOff();
+                    break;
+                }
+                case BBIO_NFC_RF_ON: {
+                    Trf797xTurnRfOn();
+                    break;
+                }
+
+                case BBIO_NFC_CMD_SEND_BITS: {
+                    chnRead(con->sdu, rx_data, 2);
+                    rlen = Trf797x_transceive_bits(rx_data[0],
+                                                   rx_data[1],
+                                                   rx_data,
+                                                   0xFF,
+                                                   10, /* 10ms TX/RX Timeout */
+                                                     0); /* TX CRC disabled */
+                    cprint(con, (char *) &rlen, 1);
+                    cprint(con, (char *) rx_data, rlen);
+                    break;
+                }
+                case BBIO_NFC_CMD_SEND_BYTES: {
+                    chnRead(con->sdu, &clen, 1);
+                    chnRead(con->sdu, rx_data, clen);
```

Comment:
Maybe add a third call to `chnRead()` to read the CRC value in a dedicated variable.
That way, `clen` contains the actual data length instead of `length+1`

Code would also be more easy to understand